sound/soc/sdca/sdca_functions.c

```diff
@@ -170,6 +171,84 @@ void sdca_lookup_functions(struct sdw_slave *slave)
 }
 EXPORT_SYMBOL_NS(sdca_lookup_functions, SND_SOC_SDCA);
 
+static int find_sdca_entity_controls(struct device *dev,
+				     struct fwnode_handle *entity_node,
+				     struct sdca_entity *entity,
+				     int func_id)
+{
+	struct fwnode_handle *control_node;
+	char control_property[40];
+	unsigned long list;
+	u32 clist = 0;
+	int i, j = 0;
+
+	if (fwnode_property_read_u32(entity_node, "mipi-sdca-control-list", &clist))
+		return 0;
+
+	entity->num_controls = hweight32(clist);
+	entity->controls = devm_kcalloc(dev, entity->num_controls,
+					sizeof(*entity->controls), GFP_KERNEL);
+	if (!entity->controls)
+		return -ENOMEM;
+
+	list = clist;
+
+	for_each_set_bit(i, &list, 32) {
+		/* DisCo uses upper-case for hex numbers */
+		snprintf(control_property, sizeof(control_property),
+			 "mipi-sdca-control-0x%X-subproperties",
+			 i);
+
+		control_node = fwnode_get_named_child_node(entity_node,
+							   control_property);
+		if (!control_node) {
+			dev_err(dev, "%s: %pfwP: property %s not found\n",
+				__func__, entity_node, control_property);
+			//FIXME: return -EINVAL;
+			continue;
+		}
+
+		if (!fwnode_property_read_u32(control_node,
+					      "mipi-sdca-control-access-mode",
+					      &clist))
+			entity->controls[j].mode = clist;
+
+		if (entity->controls[j].mode == SDCA_CONTROL_ACCESS_MODE_DC) {
+			if (!fwnode_property_read_u32(control_node,
+						      "mipi-sdca-control-dc-value",
+						      &clist))
+				entity->controls[j].value = clist;
```

Comment:
Yeah mostly this is still fairly early stage stuff for a lot of things we should return a few more errors etc., although some of it might require some thinking as I have a feeling errors might not be that rare and we generally don't get to fix the ACPI, but at least some bit fat warnings would be nice.